DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky_inversion/src/cholesky_inversion.hpp

```diff
@@ -0,0 +1,132 @@
+class IPipe;
+
+/*
+  Implementation of the Cholesky-based inversion using streaming kernels
+  Can be configured by datatype, matrix size (must use square matrices), real
+  and complex.
+*/
+template <unsigned dimension,  // Number of columns/rows in the input matrix
+          unsigned raw_latency_decomposition,  // RAW latency for triangular
+                                               // loop optimization
+          unsigned raw_latency_inversion,  // RAW latency for triangular loop
+                                           // optimization
+          bool is_complex,  // Selects between ac_complex<T> and T datatype
+          typename T,       // The datatype for the computation
+          typename TT = std::conditional_t<is_complex, ac_complex<T>, T>
+          // TT will be ac_complex<T> or T depending on is_complex
+          >
+void CholeskyInversionImpl(
+    std::vector<TT> &a_matrix,  // Input matrix A to invert
+    std::vector<TT> &i_matrix,  // Output matrix I (inverse of A)
+    sycl::queue &q,             // Device queue
+    int matrix_count,           // Number of matrices to invert
+    int repetitions             // Number of repetitions, for performance
+                                // evaluation
+) {
+  constexpr int kAMatrixSize = dimension * dimension;
+  constexpr int kIMatrixSize = dimension * (dimension + 1) / 2;
+  constexpr int kNumElementsPerDDRBurst = is_complex ? 4 : 8;
+
+  using PipeType = fpga_tools::NTuple<TT, kNumElementsPerDDRBurst>;
+
+  // Pipes to communicate the A and L matrices between kernels
+  using AMatrixPipe = sycl::ext::intel::pipe<APipe, PipeType, 3>;
+  using LMatrixPipe =
+      sycl::ext::intel::pipe<LPipe, TT, kNumElementsPerDDRBurst * 4>;
+  using IMatrixPipe =
+      sycl::ext::intel::pipe<IPipe, TT, kNumElementsPerDDRBurst * 4>;
+
+  // Allocate FPGA DDR memory.
+  TT *a_device = sycl::malloc_device<TT>(kAMatrixSize * matrix_count, q);
+  TT *i_device = sycl::malloc_device<TT>(kIMatrixSize * matrix_count, q);
+
+  if ((a_device == nullptr) || (i_device == nullptr)) {
+    std::cerr << "Error when allocating FPGA DDR" << std::endl;
+    std::cerr << "The FPGA DDR may be full" << std::endl;
+    std::cerr << "Try reducing the matrix sizes/count" << std::endl;
+    return;
+  }
+
+  // Copy the matrices to decompose to the FPGA DDR
+  q.memcpy(a_device, a_matrix.data(), kAMatrixSize * matrix_count * sizeof(TT))
+      .wait();
+
+  // Launch a kernel that will repeatedly read the matrices from the FPGA DDR
+  // and write their content to the AMatrixPipe pipe.
+  auto ddr_read_event = q.single_task<CholeskyDDRToLocalMem>([=] {
+    MatrixReadFromDDRToPipe<TT, dimension, dimension, kNumElementsPerDDRBurst,
+                            AMatrixPipe>(a_device, matrix_count, repetitions);
+  });
+
+  // Read the A matrix from the AMatrixPipe pipe and compute the Cholesky
+  // decomposition. Write the L output matrix to the LMatrixPipe pipe.
+  q.single_task<DecompositionKernel>(
```

Comment:
In an IP Authoring world, we would use task sequences here instead of multiple discrete kernels for parallelism. Perhaps someday they will work :)